src/transformers/models/whisper/generation_whisper.py

```diff
@@ -616,6 +624,23 @@ def generate(
             generation_config=generation_config,
         )
 
+        # 5bis speculative decoding: patch generate to make sure the assistant model returns the input tokens
+        if "assistant_model" in kwargs:
+            from .modeling_whisper import WhisperForCausalLM
+
+            assistant_model = kwargs["assistant_model"]
+            if assistant_model is self:
+                assistant_model = copy.deepcopy(kwargs["assistant_model"])
+            if (
+                not hasattr(assistant_model, "_generate_with_input_tokens")
+                and type(assistant_model) != WhisperForCausalLM
+            ):
+                assistant_model.generate = types.MethodType(
+                    self.generate_add_decoder_input_ids(assistant_model.generate.__func__), assistant_model
```

Comment:
# we need to return decoder input ids to make sure assistant generated tokens are the same as the shape of the logits from the forward path

or something to explain why we need to do this! 🤗 
Also really not a fan of non explicit solutions like this one, tho I am guessing there are not many other things you can do!

Reply:
added in [c9b36c3](https://github.com/huggingface/transformers/pull/34135/commits/c9b36c3a5a299f5a5f7fa9c3eadeef0f04b417d5)